acme/client.go

```diff
@@ -56,9 +61,12 @@ type Client struct {
 // the ACME directory located at caDirURL for the rest of its actions. It will
 // generate private keys for certificates of size keyBits. And, if the challenge
 // type requires it, the client will open a port at optPort to solve the challenge.
-// If optPort is blank, the port required by the spec will be used, but you must
-// forward the required port to optPort for the challenge to succeed.
-func NewClient(caDirURL string, user User, keyBits int, optPort string) (*Client, error) {
+//
+// If optSolvers is nil, the value of DefaultSolvers is used. If given explicitly,
+// it is a set of solver names to enable. The "http-01" and "tls-sni-01" solvers
+// take an optional TCP port to listen on after a colon, e.g. "http-01:80". If
```

Comment:
Agreed. The problem I'm trying to get around is that not all solvers will have the concept of a port. E.g. the file challenge for http-01 I've made has a root path, but no port. (It's a directory where Apache serves the file from.)

So the options need to be per-solver, or per category of solvers. This was the simplest solution. Another would be to make solver an exported type and allow feeding that into NewClient. But that would also require another API change (not having to supply the jws object on construction). I felt that was overkill for my purposes: allowing multiple implementations of the http-01 solver.